compiler/rustc_index/src/bit_set.rs

```diff
@@ -328,6 +342,403 @@ impl<T: Idx> BitRelations<BitSet<T>> for BitSet<T> {
+        let chunk = &self.chunks[chunk_index(elem)];
+        match &chunk {
+            Zeros(_) => false,
+            Ones(_) => true,
+            Mixed(_, _, words) => {
+                let (word_index, mask) = chunk_word_index_and_mask(elem);
+                (words[word_index] & mask) != 0
+            }
+        }
+    }
+
+    /// Insert `elem`. Returns whether the set has changed.
+    pub fn insert(&mut self, elem: T) -> bool {
+        self.sane();
+        assert!(elem.index() < self.domain_size);
+        let chunk_index = chunk_index(elem);
+        let chunk = &mut self.chunks[chunk_index];
+        match *chunk {
+            Zeros(chunk_domain_size) => {
+                if chunk_domain_size > 1 {
+                    let mut words = [0; CHUNK_WORDS];
+                    let (word_index, mask) = chunk_word_index_and_mask(elem);
+                    words[word_index] |= mask;
+                    *chunk = Mixed(chunk_domain_size, 1, Rc::new(words));
+                } else {
+                    *chunk = Ones(chunk_domain_size);
+                }
+                self.sane();
+                true
+            }
+            Ones(_) => false,
+            Mixed(chunk_domain_size, ref mut count, ref mut words) => {
+                // We skip all the work if the bit is already set.
+                let (word_index, mask) = chunk_word_index_and_mask(elem);
+                let changed = if (words[word_index] & mask) == 0 {
+                    let words = Rc::make_mut(words);
+                    words[word_index] |= mask;
+                    *count += 1;
+                    if *count == chunk_domain_size {
+                        *chunk = Ones(chunk_domain_size);
+                    }
+                    true
+                } else {
+                    false
+                };
+                self.sane();
+                changed
+            }
+        }
+    }
+
+    /// Sets all bits to true.
+    pub fn insert_all(&mut self) {
+        self.sane();
+        for chunk in self.chunks.iter_mut() {
+            *chunk = match *chunk {
+                Zeros(chunk_domain_size)
+                | Ones(chunk_domain_size)
+                | Mixed(chunk_domain_size, ..) => Ones(chunk_domain_size),
+            }
+        }
+        self.sane();
+    }
+
+    /// Returns `true` if the set has changed.
+    pub fn remove(&mut self, elem: T) -> bool {
+        self.sane();
+        assert!(elem.index() < self.domain_size);
+        let chunk_index = chunk_index(elem);
+        let chunk = &mut self.chunks[chunk_index];
+        match *chunk {
+            Zeros(_) => false,
+            Ones(chunk_domain_size) => {
+                if chunk_domain_size > 1 {
+                    // Set only the bits in use.
+                    let num_words = num_words(chunk_domain_size as usize);
+                    let mut words = [0; CHUNK_WORDS];
+                    words[..num_words].fill(!0);
```

Comment:
nit: it's probably better to initialize with `[!0; CHUNK_WORDS]` -- or we might borrow from my earlier comment about initializing in-place inside the Rc.

In a toy example at least I'm seeing the zeroing and a subsequent fill look pretty unfortunate in terms of generated assembly, particularly when we're going to end up clearing the excess bits regardless.

Not sure how much it'll matter in practice.